services/mqtt/service.go

```diff
@@ -233,11 +242,81 @@ type handler struct {
 
 func (h *handler) Handle(event alert.Event) {
 	h.diag.HandlingEvent()
-	if err := h.s.Alert(h.c.BrokerName, h.c.Topic, h.c.QoS, h.c.Retained, event.State.Message); err != nil {
+	topic, err := h.renderTopic(h.c.Topic, event)
+	if err != nil {
+		h.diag.Error("failed to create MQTT topic from template", err)
+	} else if err = h.s.Alert(h.c.BrokerName, topic, h.c.QoS, h.c.Retained, event.State.Message); err != nil {
 		h.diag.Error("failed to post message to MQTT broker", err)
 	}
 }
 
+type idInfo struct {
+	// Measurement name
+	Name string
+
+	// Task name
+	TaskName string
+
+	// Concatenation of all group-by tags of the form [key=value,]+.
+	// If not groupBy is performed equal to literal 'nil'
+	Group string
+
+	// Map of tags
+	Tags map[string]string
+}
+
+type messageInfo struct {
+	idInfo
+
+	// The ID of the alert.
+	ID string
+
+	// Fields of alerting data point.
+	Fields map[string]interface{}
+
+	// Alert Level, one of: INFO, WARNING, CRITICAL.
+	Level string
+
+	// Time
+	Time time.Time
+
+	// Duration of the alert
+	Duration time.Duration
+}
+
+func (h *handler) renderTopic(topic string, event alert.Event) (string, error) {
+	minfo := messageInfo{
+		idInfo: idInfo{
+			Name:     event.Data.Name,
+			TaskName: event.Data.TaskName,
+			Group:    event.Data.Group,
+			Tags:     event.Data.Tags,
+		},
+		ID:       event.State.ID,
+		Fields:   event.Data.Fields,
+		Level:    event.State.Level.String(),
+		Time:     event.State.Time,
+		Duration: event.State.Duration,
+	}
+
+	topicBuf := h.s.bufPool.Get().(*bytes.Buffer)
+	defer func() {
+		topicBuf.Reset()
+		h.s.bufPool.Put(topicBuf)
+	}()
+
+	topicTmpl, err := text.New("topic").Parse(topic)
```

Comment:
The topic template doesn't change for a given handler, can you udpate this to parse the template during handler construction? Then you can eval the template for each eval without parsing it. This also provides the user with earlier feedback if they have a typo in their template.

Reply:
Done.